includes/class-wcs-action-scheduler-customer-notifications.php

```diff
@@ -0,0 +1,290 @@
+ */
+class WCS_Action_Scheduler_Customer_Notifications extends WCS_Scheduler {
+
+	/**
+	 * @var int Time offset (in whole hours) between the notification and the action it's notifying about.
+	 */
+	protected int $hours_offset;
+
+	/**
+	 * @var array|string[] Notifications scheduled by this class.
+	 *
+	 * Just for reference.
+	 */
+	protected array $notification_actions = array(
+		'woocommerce_scheduled_subscription_customer_notification_trial_expiration',
+		'woocommerce_scheduled_subscription_customer_notification_expiration',
+		'woocommerce_scheduled_subscription_customer_notification_manual_renewal',
+		'woocommerce_scheduled_subscription_customer_notification_auto_renewal',
+	);
+
+	public function get_hours_offset( $subscription ) {
+		/**
+		 * Offset between a subscription event and related notification.
+		 *
+		 * @since 8.0.0
+		 *
+		 * @param int $hours_offset
+		 */
+		return apply_filters( 'woocommerce_subscriptions_customer_notification_hours_offset', $this->hours_offset, $subscription );
+	}
+
+	public function set_hours_offset( $hours_offset ) {
+		$this->hours_offset = $hours_offset;
+	}
+
+	public function __construct() {
+		parent::__construct();
+
+		$setting_option     = get_option(
+			WC_Subscriptions_Admin::$option_prefix . WC_Subscriptions_Email_Notifications::$offset_setting_string,
+			array(
+				'number' => 3,
+				'unit'   => 'days',
+			)
+		);
+		$this->hours_offset = self::convert_offset_to_hours( $setting_option );
+	}
+
+	/**
+	 * Calculate time offset in hours from the settings array.
+	 *
+	 * @param array $offset Format: array( 'number' => 3, 'unit' => 'days' )
+	 *
+	 * @return float|int
+	 */
+	protected static function convert_offset_to_hours( $offset ) {
+		$default_offset = 3 * DAY_IN_SECONDS / HOUR_IN_SECONDS;
+
+		if ( ! isset( $offset['unit'] ) || ! isset( $offset['number'] ) ) {
+			return $default_offset;
+		}
+
+		switch ( $offset['unit'] ) {
+			case 'days':
+				return ( $offset['number'] * DAY_IN_SECONDS / HOUR_IN_SECONDS );
+			case 'weeks':
+				return ( $offset['number'] * WEEK_IN_SECONDS / HOUR_IN_SECONDS );
+			case 'months':
+				return ( $offset['number'] * MONTH_IN_SECONDS / HOUR_IN_SECONDS );
+			case 'years':
+				return ( $offset['number'] * YEAR_IN_SECONDS / HOUR_IN_SECONDS );
+			default:
+				return $default_offset;
+		}
+	}
+
+	public function set_date_types_to_schedule() {
+		$date_types_to_schedule = wcs_get_subscription_date_types();
+		unset(
+			$date_types_to_schedule['start'],
+			$date_types_to_schedule['cancelled'] // prevent scheduling end date when reactivating subscription.
+		);
+
+		//TODO: filter?
+		$this->date_types_to_schedule = array_keys( $date_types_to_schedule );
+	}
+
+	protected function schedule_notification( $subscription, $action, $timestamp ) {
+		if (
+			! (
+				$subscription->has_status( 'active' )
+				|| $subscription->has_status( 'pending-cancel' ) //TODO: do we want to create notifications when user cancelled the subscription?
```

Comment:
> do we want to create notifications when user cancelled the subscription?

I think we'll need to test that to confirm but I don't think we would.

For some background the `WCS_Scheduler` hooks into both subscription date and status updates. 

When a subscription date is updated, we schedule the related AS event. 
When a subscription status is changed, we unschedule any events that are no longer applicable to that status and schedule events that are. eg putting a subscription on-hold cancels the payment event. Reactivating a subscription reschedules the next payment date.

I think this notification class should follow a similar principle.

So to answer the question then I don't think we do need to create notifications on the cancelled event because when a customer cancels the subscriptions, we calculate the end date and schedule the end date. The scheduling of that date is what should trigger the notification to be scheduled, not the status transition itself.

Reply:
Thanks for elaborating. My todo comment was added to clarify whether we want to notify the customer about the upcoming expiry if they manually canceled the subscription (and thus should be aware that it's expiring).

Reply:
Agreed!